impeller/fixtures/impeller.frag

```diff
@@ -0,0 +1,436 @@
+                float shadow_multiplier,
+                vec4 ray_noise) {
+  vec4 result_color;
+  if (dist >= kMaxDistance) {
+    result_color = EnvironmentColor(ray_direction);
+  } else {
+    vec3 surface_position = ray_position + ray_direction * dist;
+    result_color = SurfaceColor(ray_direction, surface_position, surface_normal,
+                                material, shadow_multiplier);
+  }
+  float glow_factor = float(steps_taken) / float(kMaxSteps);
+  vec4 glow_color =
+      mix(kGlowColor, kGlowColor2, sin(frame_info.time / 3.0) * 0.5 + 0.5);
+  return mix(result_color, glow_color, glow_factor * kGlowBlend);
+}
+
+vec4 CombinedColor(vec3 ray_position, vec3 ray_direction, vec4 ray_noise) {
+  int steps_taken;
+  vec2 result = March(ray_position, ray_direction, steps_taken, false, false);
+  ray_position = ray_position + ray_direction * result.x;
+  vec3 surface_normal = SceneGradient(ray_position);
+
+  float glass_reflection_factor = 0.0;
+  vec4 glass_reflection_color = vec4(0);
+  if (result.y == -2.0) {  // March into the glass.
+    vec3 glass_reflection_direction = reflect(ray_direction, surface_normal);
+    glass_reflection_color = EnvironmentColor(glass_reflection_direction);
+    glass_reflection_factor =
+        0.5 - dot(glass_reflection_direction, surface_normal) * 0.6;
+
+    ray_direction = refract(ray_direction, surface_normal, kGlassIOR);
+    ray_position += ray_direction * 0.5;
+    int steps;
+    result = March(ray_position, ray_direction, steps, true, false);
+    steps_taken += steps;
+
+    ray_position = ray_position + ray_direction * result.x;
+    surface_normal = InnerGlassGradient(ray_position);
+  }
+
+  if (result.y == -3.0) {  // March out of the glass.
+    ray_direction = refract(ray_direction, surface_normal, 1.0 / kGlassIOR);
+    ray_position += ray_direction * 0.5;
+    int steps;
+    result = March(ray_position + ray_direction * result.x, ray_direction,
+                   steps, false, false);
+    steps_taken += steps;
+    ray_position = ray_position + ray_direction * result.x;
+    surface_normal = InnerGlassGradient(ray_position);
+  }
+
+  float shadow_multiplier = MarchShadow(ray_position);
+  vec4 scene_color =
+      SceneColor(ray_position, ray_direction, surface_normal, result.x,
+                 result.y, steps_taken, shadow_multiplier, ray_noise);
+
+  return mix(scene_color, glass_reflection_color, glass_reflection_factor);
+}
+
+//------------------------------------------------------------------------------
+/// Camera/lens.
+///
+
+vec3 GetFragDirection(vec2 uv, vec3 cam_forward) {
+  vec2 lens_uv =
+      (uv - 0.5 * frame_info.texture_size) / frame_info.texture_size.xx;
+  vec3 cam_right = cross(cam_forward, vec3(0, 1, 0));
+  vec3 cam_up = cross(cam_forward, cam_right);
+
+  float fov = 65.0 * kPi / 180.0;
+  return normalize(cam_forward * cos(fov) + cam_right * lens_uv.x * sin(fov) +
+                   cam_up * lens_uv.y * sin(fov));
+}
+
+void main() {
```

Comment:
I tried to profile the performance of this kernel in the shader profiler but it only seems to be available on iOS and TVOS (and perhaps the M1 Macs)? It would be interesting to take a stab at optimizing this using a profiler. I bet there are many low hanging fruit.

Reply:
Oh definitely, this can get way faster. One particularly egregious example is that the rotations should be computed once, but they're computed during distance evaluation time (literally 300-400x+ more expensive than necessary).